### PR TITLE
feat(plugin-chart-echarts): implement event interval and timeseries annotations

### PR DESCRIPTION
**Pull request by villebro**
Add support for event, interval and time series annotations to ECharts Timeseries chart.

image

### CODE REVIEW

File: packages/superset-ui-core/src/query/types/AnnotationLayer.ts [hunk 1]
```diff
@@ -54,6 +56,11 @@ export type IntervalAnnotationLayer = BaseAnnotationLayer &
     annotationType: 'INTERVAL';
   };
 
+export type TableAnnotationLayer = BaseAnnotationLayer &
+  TableSourceAnnotationLayer & {
+    annotationType: 'EVENT' | 'INTERVAL';
```
Comment (ktmud) [inline packages/superset-ui-core/src/query/types/AnnotationLayer.ts:61]: Can we give `annotationType` its own type? Either a string union literal type or `Enum` would be fine.

```ts
type AnnotationType = 'Event' | 'Interval' | ...

interface BaseAnnotationLayer {
  annotationType: AnnotationType
}
```

File: plugins/plugin-chart-echarts/src/Timeseries/transformProps.ts [hunk 2]
```diff
@@ -127,24 +144,140 @@ export default function transformProps(chartProps: ChartProps): EchartsTimeserie
...
-          type: style,
-          width: annotationWidth,
-        },
-        type: 'line',
-        smooth: true,
-        // @ts-ignore
-        data: evalFormula(layer, data),
-        symbolSize: 0,
-      });
+    if (annotationShow) {
```
Comment (ktmud) [inline plugins/plugin-chart-echarts/src/Timeseries/transformProps.ts:147]: These if branches (and the transform function itself) are growing a little bit too big, do you think if it makes sense to split some pieces into smaller functions? For example, maybe each annotation transformer can have its own function?

**Review by ktmud — COMMENTED** [review]
Super useful feature! Glad to see the ECharts-powered visualization becomes more and more powerful. Added a couple of comments on typing and code structure.

File: packages/superset-ui-core/src/query/types/AnnotationLayer.ts (hunk 1 above)
Comment (villebro) [reply packages/superset-ui-core/src/query/types/AnnotationLayer.ts:61]: Good idea

File: plugins/plugin-chart-echarts/src/Timeseries/transformProps.ts (hunk 2 above)
Comment (villebro) [reply plugins/plugin-chart-echarts/src/Timeseries/transformProps.ts:147]: You're right, this was probably overdue, will split it up.

**Comment by villebro** [comment]
@ktmud I've split up the series and annotation transformers out of the main loop and replaced the enum-like fields with enums. There's still potential for more refactoring, but this feels like an ok first step that can be improved upon in subsequent PRs (there will undoubtedly be follow up PRs to this).

**Comment by villebro** [comment]
@ktmud it would be great to get a re-review of this PR so we can introduce this functionality.

**Review by ktmud — APPROVED** [review]
LGTM! Sorry for the delay. Totally missed the notification on this one.

**Pull request merged**